src/Expr.cc

```diff
@@ -116,6 +116,7 @@ int Expr::num_exprs = 0;
 Expr::Expr(ExprTag arg_tag) : tag(arg_tag), paren(false), type(nullptr) {
     SetLocationInfo(&start_location, &end_location);
     opt_info = new ExprOptInfo();
+    register_new_expr(ThisPtr());
```

Comment:
The pinning only during main parsing is a good thought - that would keep things under our control. We could also extend `ThisPtr()` to work on `const` Exprs/Stmts, in which case when adding bare pointers to containers the AST analysis could also stash an IntrusivePtr version, rather than doing it wholesale for every possible AST node.

Reply:
> The pinning only during main parsing is a good thought

I actually meant that the pinning could be local to script_opt via an extra traversal. Just before doing the "critical" stuff, have a guard/scope class that keeps the AST objects alive until destruction.
```
+    class PinAstObjs : public TraversalCallback {
+    public:
+        PinAstObjs() { zeek::detail::traverse_all(this); };
+        TraversalCode PreStmt(const Stmt* stmt) override {
+            ast_objs.push_back(const_cast<Stmt*>(stmt)->ThisPtr());
+            return TC_CONTINUE;
+        }
+
+        TraversalCode PreExpr(const Expr* expr) override {
+            ast_objs.push_back(const_cast<Expr*>(expr)->ThisPtr());
+            return TC_CONTINUE;
+        }
+
+        std::vector<zeek::IntrusivePtr<Obj>> ast_objs;
+    };
+
+    PinAstObjs pin_ast_objs;
+
```

That costs an extra traversal, but doubt that matters too much. Also suspect ZAM is traversing the AST anyhow, so could also do it there. New hooks in Stmt/Expr constructors stashing away pointers isn't great.

Reply:
I was thinking further about this and I believe it's actually hazardous. The problem is that during script optimization the AST evolves in an ongoing fashion, so new nodes get constructed and then potentially discarded. The danger is that a pointer to one of those winds up in a container, and it won't be caught by a particular AST traversal since it popped up in the flow of optimization subsequent to the traversal.

This has me thinking that a `ThisPtr()` that works for `const` objects is the best way to go. Here's a thought on how perhaps to do that safely: add an `Obj` wrapper class that will hold an `IntrusivePtr<Obj>` inside it, constructed from a `const Obj*`, but won't allow any access to the pointer. That way, there's no temptation to sneak around `const`-ness but we still address the pointer mis-aliasing issue.  What do you think?